src/index.ts

```diff
@@ -10,22 +34,400 @@ import {list} from 'white-space-x';
+
+  return processedChars.toString('utf-8', startIndex, endIndex);
+}
+
+export function normalizeSpaces10(string: string): string {
+  const chars = Buffer.from(string);
+  const processedChars = Buffer.alloc(chars.length, SPACE_CODE);
+
+  let lastProcessedIndex = -1;
+  let isLastCharWhitespace = 0;
+
+  for (let i = 0; i < chars.length; i++) {
+    const char = chars[i];
+    const isWhitespace = isWhiteSpaceCharCodeNumber(char);
+
+    if (!isWhitespace) {
+      lastProcessedIndex++;
+
+      processedChars[lastProcessedIndex] = char;
+    } else {
+      if (isLastCharWhitespace) {
+        continue;
+      }
+
+      lastProcessedIndex++;
+    }
+
+    isLastCharWhitespace = isWhitespace;
+  }
+
+  // If the original string has whitespaces at the start or at the end, the first and the
+  // last processed chars should be single whitespaces, so we need to make offset from both
+  // ends to trim they
+  const firstChar = processedChars[0];
+  const lastChar = processedChars[lastProcessedIndex];
+
+  const startIndex = firstChar === SPACE_CODE ? 1 : 0;
+  const endIndex = lastChar === SPACE_CODE ? lastProcessedIndex : lastProcessedIndex + 1;
+
+  return processedChars.toString('utf-8', startIndex, endIndex);
+}
+
+export function normalizeSpaces11(string: string): string {
+  const chars = Buffer.from(string);
+  const processedChars = Buffer.alloc(chars.length, SPACE_CODE);
+
+  let lastProcessedIndex = -1;
+  let isLastCharWhitespace = 0;
+
+  for (let i = 0; i < chars.length; i++) {
+    const char = chars[i];
+    const isWhitespace = isWhiteSpaceCharCodeIndexedUint8Array(char);
+
+    if (!isWhitespace) {
+      lastProcessedIndex++;
+
+      processedChars[lastProcessedIndex] = char;
+    } else {
+      if (isLastCharWhitespace) {
+        continue;
+      }
+
+      lastProcessedIndex++;
+    }
+
+    isLastCharWhitespace = isWhitespace;
+  }
+
+  // If the original string has whitespaces at the start or at the end, the first and the
+  // last processed chars should be single whitespaces, so we need to make offset from both
+  // ends to trim they
+  const firstChar = processedChars[0];
+  const lastChar = processedChars[lastProcessedIndex];
+
+  const startIndex = firstChar === SPACE_CODE ? 1 : 0;
+  const endIndex = lastChar === SPACE_CODE ? lastProcessedIndex : lastProcessedIndex + 1;
+
+  return processedChars.toString('utf-8', startIndex, endIndex);
+}
+
+export function normalizeSpaces12(text: string): string {
+  const res = execSync(`tr -s "[:space:]" | tr "[:space:][:space:]" " " | tr -s "[:space:]"`, {
+    maxBuffer: 999999999999,
+    input: text,
+  });
+
+  return res.toString('utf-8').trim();
+}
+
+export async function normalizeSpaces13(string: string): Promise<string> {
+  if (string.length < 3145728) {
+    return normalizeSpaces11(string);
+  }
+
+  const chars = Buffer.from(string);
+  const charsPart1 = chars.slice(0, chars.length / 2);
+  const charsPart2 = chars.slice(charsPart1.length);
```

Comment:
@vladgolubev @mpushkin  What do you think if we would parallel execution, not for a fixed number(2) but depending on the current host's CPU info using `os.cpu().length` from `os` module that returns a number of logical CPUs?

Reply:
I tested with 4 threads and it was slower for me, looks like the overhead of creating a new worker is bigger than the benefit. 

I think more threads will be better for humongous files like 3 GB strings, but for 33 MB it was slower

But I like the idea, it is the correct one, instead of the guesstimate. We can cleverly define how many threads to use depending on the string size

Reply:
However, it is definitely better to rewrite the current realization for working with a customs number of threads(and specify 2 as the default param value). In this way, we can delegate choosing the concurrency level to the callee code.